plugins/out_skywalking/skywalking.c

```diff
@@ -0,0 +1,387 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_output_plugin.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_time.h>
+#include <fluent-bit/flb_http_client.h>
+
+#include "skywalking.h"
+
+#define DEFAULT_SW_OAP_HOST "127.0.0.1"
+#define DEFAULT_SW_OAP_PORT 12800
+#define DEFAULT_SW_SVC_NAME "sw-service"
+#define DEFAULT_SW_INS_NAME "fluent-bit"
+#define DEFAULT_SW_LOG_PATH "/v3/logs"
+
+static int cb_sw_init(struct flb_output_instance *ins, struct flb_config *config, void *data) {
+    int io_flags;
+    struct flb_output_sw *ctx;
+    const char* tmp;
+
+    /* Allocate plugin context */
+    ctx = flb_calloc(1, sizeof(struct flb_output_sw));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+    ctx->ins = ins;
+
+    /* setup access token */
+    tmp = flb_output_get_property("auth_token", ins);
+    if (tmp) {
+        flb_plg_debug(ctx->ins, "OAP token: %s", ctx->auth_token);
+        ctx->auth_token = flb_sds_create(tmp);
+    }
+
+    /* setup uri */
+    ctx->uri = flb_sds_create(DEFAULT_SW_LOG_PATH);
+    if (!ctx->uri) {
+        flb_plg_error(ctx->ins, "failed to configure endpoint");
+        flb_free(ctx);
+        return -1;
+    }
+
+    /* configure service/instance name */
+    tmp = flb_output_get_property("service_name", ins);
+    if (tmp) {
+        ctx->svc_name = flb_sds_create(tmp);
+    } else {
+        ctx->svc_name = flb_sds_create(DEFAULT_SW_SVC_NAME);
+    }
+
+    tmp = flb_output_get_property("instance_name", ins);
+    if (tmp) {
+        ctx->svc_inst_name = flb_sds_create(tmp);
+    } else {
+        ctx->svc_inst_name = flb_sds_create(DEFAULT_SW_INS_NAME);
+    }
+
+    flb_plg_debug(ctx->ins, "configured %s/%s", ctx->svc_name, ctx->svc_inst_name);
+
+    flb_output_net_default(DEFAULT_SW_OAP_HOST, DEFAULT_SW_OAP_PORT, ins);
+
+    /* scheme configuration */
+    if (ins->use_tls == FLB_TRUE) {
+        io_flags = FLB_IO_TLS;
+        ctx->http_scheme = flb_sds_create("https://");
+    } else {
+        io_flags = FLB_IO_TCP;
+        ctx->http_scheme = flb_sds_create("http://");
+    }
+
+    /* setup host and port */
+    ctx->host = ins->host.name;
+    ctx->port = ins->host.port;
+    flb_plg_debug(ctx->ins, "OAP address: %s:%d", ctx->host, ctx->port);
+
+    /* configure upstream instance */
+    ctx->u = flb_upstream_create(config, ctx->host, ctx->port, io_flags, ins->tls);
+    if (!ctx->u) {
+        flb_plg_error(ctx->ins, "failed to create upstream context");
+        flb_free(ctx);
```

Comment:
you are leaking all the memory allocated previously, it's better to create a destroy routine for your context